File: task4.py
import sys


class Node:
    def __init__(self, x):
        self.val = x
        self.next = None

    def print(self):
        cur = self
        while cur:
            print(cur.val, end=' ')
            cur = cur.next
        print()


def split_list(head):
    end_finder = head
    middle_finder = head

    while end_finder and end_finder.next:
        middle_finder = middle_finder.next
        end_finder = end_finder.next.next

    middle = middle_finder.next
    middle_finder.next = None

    return head, middle


def reverse(head):
    prev = None
    current = head

    while current:
        next = current.next
        current.next = prev
        prev = current
        current = next

    return prev


def merge(first, second):
    result = first
    head = first
    first = first.next

    while second:
        result.next = second
        result = result.next
        second = second.next
        if first:
            first, second = second, first

    return head


def create_list(values_list):
    try:
        head = Node(int(values_list[0]))
        cur = head
        for val in values_list[1:]:
            cur.next = Node(int(val))
            cur = cur.next
    except Exception as e:
        print('Incorrect argument. Use python task4.py <int> [<int> ...]')
        head = None
    return head


def reorder_list(head):
    first, second = split_list(head)
    second = reverse(second)
    result = merge(first, second)

    return result


lst = create_list(sys.argv[1:])
if lst is not None:
    reorder_list(lst).print()

File: task4.cpp
#include <iostream>

struct Node {
        int value;
        Node *next = nullptr;
        Node(int x) {
                this->value = x;
        }
};

Node *create_list(int argc, char const *argv[]) {
        if (argc < 2)
                return nullptr;
        Node *first_node = new Node(atoi(argv[1])), *cur = first_node;
        for (size_t i = 2; i < argc; ++i) {
                cur->next = new Node(atoi(argv[i]));
                cur = cur->next;
        }
        return first_node;
}

void print_list(Node *head) {
        while (head) {
                std::cout << head->value << ' ';
                head = head->next;
        }
        std::cout << '\n';
}

Node *split_list(Node *head) {
        Node *end_finder = head, *prev = head;
        while (end_finder && end_finder->next) {
                prev = head;
                head = head->next;
                end_finder = end_finder->next->next;
        }
        if (end_finder) {
                prev = head;
                head = head->next;
        }
        prev->next = nullptr;
        return head;
}

Node *reverse_list(Node *head) {
        Node *tmp = nullptr;
        Node *prev = nullptr;
        while (head) {
                tmp = head->next;
                head->next = prev;
                prev = head;
                head = tmp;
        }
        return prev;
}

Node *merge(Node *first, Node *second) {
        Node *result = first, *head = result;
        first = first->next;
        while (second && first) {
                result->next = second;
                result = result->next;
                second = second->next;
                result->next = first;
                result = result->next;
                first = first->next;
        }
        if (second) {
                result->next = second;
                result = result->next;
        }
        return head;
}

Node *reorder_list(Node *head) {
        Node *mid = split_list(head);
        mid = reverse_list(mid);
        Node *result = merge(head, mid);
        return result;
}

int main(int argc, char const *argv[]) {
        Node *head = create_list(argc, argv);
        if (head)
                print_list(reorder_list(head));
        else
                std::cout << "Incorrect usage. Use ./task4.out <int> [<int> ...]"
                          << '\n';
        return 0;
}
